engine/runtime-compiler/src/main/java/org/enso/compiler/pass/resolve/MethodDefinitions.java

```diff
@@ -0,0 +1,338 @@
+                            explicitMethod.copy(
+                                resolvedMethodRef,
+                                explicitMethod.body(),
+                                isStatic,
+                                explicitMethod.isPrivate(),
+                                explicitMethod.isStaticWrapperForInstanceMethod(),
+                                explicitMethod.location(),
+                                explicitMethod.passData(),
+                                explicitMethod.diagnostics(),
+                                explicitMethod.id());
+                        yield resolvedMethod;
+                      }
+                      case Method.Conversion conversionMethod -> {
+                        var sourceTypeExpr = conversionMethod.sourceTypeName();
+                        Name resolvedName =
+                            switch (sourceTypeExpr) {
+                              case Name name -> resolveType(name, availableSymbolsMap);
+                              default -> new Conversion(
+                                  sourceTypeExpr,
+                                  UnsupportedSourceType$.MODULE$,
+                                  new MetadataStorage());
+                            };
+                        var resolvedMethod =
+                            conversionMethod.copy(
+                                resolvedMethodRef,
+                                resolvedName,
+                                conversionMethod.body(),
+                                conversionMethod.location(),
+                                conversionMethod.passData(),
+                                conversionMethod.diagnostics(),
+                                conversionMethod.id());
+                        yield resolvedMethod;
+                      }
+                      default -> throw new CompilerError(
+                          "Unexpected method type in MethodDefinitions pass.");
+                    };
+                  } else {
+                    return def;
+                  }
+                });
+
+    java.util.List<Definition> withStaticAliases = new ArrayList<>();
+    for (var def : CollectionConverters.asJava(newDefs)) {
+      withStaticAliases.add(def);
+      if (def instanceof Method.Explicit method && !method.isStatic()) {
+        var staticAlias = generateStaticAliasMethod(method);
+        if (staticAlias != null) {
+          withStaticAliases.add(staticAlias);
+        }
+      }
+    }
+
+    return ir.copyWithBindings(CollectionConverters.asScala(withStaticAliases).toList());
+  }
+
+  /**
+   * Returns null if there is no suitable static alias method that can be generated for the given
+   * {@code method}.
+   *
+   * @param method Non-static method from which a static alias method is generated.
+   * @return Static alias method for the given {@code method} or null.
+   */
+  private Method.Explicit generateStaticAliasMethod(Method.Explicit method) {
```

Comment:
Unrelated to this PR.

What would it take to **register only static methods** in each type? What's an invocation of `(x:Integer).to_text` anyway? It should be convertible to `Integer.to_text x` - e.g. an instance invocation just resolves the right method based on the type of the instances, but then it invokes the static `Function`.

Benefit:
- reduce the amount of `Function` instances in the system by half

Prior art:
- Graal compiler sees all JVM methods as static (instances methods of a class being static and having `this` as first argument)
- the compiler only has special `InvokeNode` that deals with finding the right method to invoke when virtual dispatch is necessary

Reply:
Created an issue https://github.com/enso-org/enso/issues/11686 from your comment.